erpnext/erpnext_integrations/doctype/leaflink_settings/leaflink_settings.py

```diff
@@ -0,0 +1,191 @@
+		frappe.throw(_("An Exception: {e} occured while trying to sync customer on LeafLink. Please try again.".format(e)))
+
+	customers_on_leaflink = requests.get(customers_url, headers=header, params={"name": customer_name}).json()
+
+	#create customer record on leaflink if it does not exist
+	if not customers_on_leaflink.get("count"):
+
+		#create new customer data
+		new_customer = {
+			"name": customer_name,
+			"owner": leaflink_settings.get("leaflink_id")
+	 	}
+
+		#post request to create customer in leaflink
+		try:
+			post_response = requests.post(customers_url, headers=header, data=new_customer)
+			if post_response.status_code == 201:
+				frappe.msgprint(_("Customer successfully created on LeafLink!"))
+		except Exception as e:
+			frappe.throw(_("An Exception: {e} occured while trying to sync customer on LeafLink. Please try again.".format(e)))
+
+	else:
+		frappe.msgprint(_("Customer already exists on LeafLink!"))
+
+
+def authenticate_leaflink_request(leaflink_settings, integration_request):
+	"""
+	Authenticate an incoming request from LeafLink before creating a sales order.
+
+	Args:
+		leaflink_settings (dict]): contains leaflink configurations stored in Bloomstack
+		integration_request (dict]): contains details of the integration request for the sale order
+	"""
+	#ensure leaflink is enabled before proceeding further.
+	if not leaflink_settings.get("enable_leaflink"):
+		integration_request.update({
+			"status": "Failed",
+			"error": "Leaflink is not enabled."
+		})
+		integration_request.save(ignore_permissions=True)
+		return
+
+	#validate incoming keys
+	key = bytes(leaflink_settings.get("incoming_api_authentication_key"), 'utf-8')
+	incoming_signature = frappe.local.request.headers['LL-Signature'].encode("utf-8")
+	request_body = frappe.local.request.data
+	computed_signature = base64.b64encode(hmac.new(key, request_body, digestmod=hashlib.sha256).digest())
+	success = computed_signature == incoming_signature
+	if not success:
+		integration_request.update({
+			"status": "Failed",
+			"error": "Signatures do not match. Please check the Incoming Key in LeafLink Settings."
+		})
+		integration_request.save(ignore_permissions=True)
+		return False
+	return True
+
+
+@frappe.whitelist(allow_guest=True)
+def sales_order_from_leaflink(**args):
+
+	#create integration request
+	integration_request = frappe.new_doc("Integration Request")
+	integration_request.update({
+		"integration_type": "Host",
+		"integration_request_service": "LeafLink",
+		"status": "Queued",
+		"data":  json.dumps(args, default=json_handler),
+		"reference_doctype": "Sales Order"
+	})
+	integration_request.save(ignore_permissions=True)
+
+	#curate data from settings for post request
+	leaflink_settings = frappe.get_cached_doc("LeafLink Settings")
+	header = {
+		leaflink_settings.get("api_key"): leaflink_settings.get("api_value")
+	}
+
+	#validate incoming keys
+	if not authenticate_leaflink_request(leaflink_settings, integration_request):
```

Comment:
Make this authentication as the first step of this function, no need to create an Integration Request if the request is not authenticated.

Reply:
As mentioned above.